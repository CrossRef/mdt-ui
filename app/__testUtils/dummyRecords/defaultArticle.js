import {fromJS} from 'immutable'

export default {
  publicationDoi: "10.5555/defaultPub",
  articleDoi: "10.5555/defaultArticle",


  publicationJson: fromJS({
    "status": "ok",
    "message-type": "get",
    "message-version": "1.0.0",
    "message": {
      "doi": "10.5555/defaultPub",
      "owner-prefix": "10.5555",
      "type": "Publication",
      "title": {"title":"test new system 2"},
      "state": {},
      "date": "2018-01-17",
      "deposit-timestamp": "20180117154727430",
      "mdt-version": "0",
      "status": "accepted",
      "content": "<Journal xmlns=\"http://www.crossref.org/xschema/1.1\"><journal_metadata language=\"af\"><full_title>test new system 2</full_title><abbrev_title>arhaerh</abbrev_title><archive_locations><archive name=\"LOCKSS\"/></archive_locations><doi_data><doi>10.5555/13g14g14h12h2h24h2</doi><resource>http://afhaehaerhaerh.com</resource></doi_data></journal_metadata></Journal>",
      "contains": [
        {
          "doi": "10.5555/afvafbafgnafnafnafgn",
          "owner-prefix": "10.5555",
          "type": "article",
          "title": {"title":"test references"},
          "state": {},
          "date": "2018-01-17",
          "deposit-timestamp": "20180117154727430",
          "mdt-version": "1",
          "status": "draft",
          "content": "",
          "contains": [

          ]
        },
        {
          "doi": "10.5555/fbaefbadbnanana",
          "owner-prefix": "10.5555",
          "type": "article",
          "title": {"title":"test deposit timestamp"},
          "state": {},
          "date": "2018-01-17",
          "deposit-timestamp": "",
          "mdt-version": "1",
          "status": "failed",
          "content": "",
          "contains": [

          ]
        },
        {
          "doi": "10.5555/abaebaebaerbaerb",
          "owner-prefix": "10.5555",
          "type": "article",
          "title": {"title":"arahhaharh"},
          "state": {},
          "date": "2018-01-19",
          "deposit-timestamp": "",
          "mdt-version": "3",
          "status": "draft",
          "content": "",
          "contains": [

          ]
        },
        {
          "doi": "10.5555/aerhaerhaerhaerh",
          "owner-prefix": "10.5555",
          "type": "article",
          "title": {"title":"awerarehaerharh"},
          "state": {},
          "date": "2018-01-17",
          "deposit-timestamp": "",
          "mdt-version": "1",
          "status": "draft",
          "content": "",
          "contains": [

          ]
        },
        {
          "doi": "10.5555/defaultArticle",
          "owner-prefix": "10.5555",
          "type": "article",
          "title": {"title":"afhaerharhtaer"},
          "state": {},
          "date": "2018-01-23",
          "deposit-timestamp": "",
          "mdt-version": "1",
          "status": "draft",
          "content": "",
          "contains": [

          ]
        }
      ]
    }
  }),


  articleJson: fromJS({
    "status": "ok",
    "message-type": "get",
    "message-version": "1.0.0",
    "message": {
      "doi": "10.5555/defaultPub",
      "owner-prefix": "10.5555",
      "type": "Publication",
      "title": {"title":"test new system 2"},
      "state": {},
      "date": "2018-01-17",
      "deposit-timestamp": "20180117154727430",
      "mdt-version": "0",
      "status": "accepted",
      "contains": [
        {
          "doi": "10.5555/defaultArticle",
          "owner-prefix": "10.5555",
          "type": "article",
          "title": {"title":"afhaerharhtaer"},
          "state": {},
          "date": "2018-01-23",
          "deposit-timestamp": "",
          "mdt-version": "1",
          "status": "draft",
          "content": "<?xml version=\"1.0\" encoding=\"UTF-8\"?><crossref xmlns=\"http://www.crossref.org/xschema/1.1\"><journal><journal_metadata language=\"af\"><full_title>test new system 2</full_title><abbrev_title>arhaerh</abbrev_title><archive_locations><archive name=\"LOCKSS\"/></archive_locations><doi_data><doi>10.5555/13g14g14h12h2h24h2</doi><resource>http://afhaehaerhaerh.com</resource></doi_data></journal_metadata><journal_article><titles><title>afhaerharhtaer</title></titles><doi_data><doi>10.5555/aefbtabangagnanatn</doi><resource>http://</resource></doi_data></journal_article></journal></crossref>",
          "contains": [

          ]
        }
      ]
    }
  })
}
