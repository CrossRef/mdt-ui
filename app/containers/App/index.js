import React, { Component } from 'react'
import Header from '../../components/App/Header'
import Footer from '../../components/App/Footer'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import Modal from '../Modal'
import { stateTrackerII } from 'my_decorators'


export class App extends Component {
  render () {
    return (
      <div className='app'>
        <div className='app-contents'>
          <Header path={this.props.path} />
          <div className='page-contents'>
            {this.props.children}
          </div>
        </div>
        <Footer />
        <Modal />
      </div>
    )
  }
}

function mapStateToProps (state) {
  return {
    path: window.location.pathname
  }
}

function mapDispatchToProps (dispatch) {
  return { }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App)