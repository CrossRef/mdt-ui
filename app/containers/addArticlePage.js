import React, { Component } from 'react'
import ReactDOM from 'react-dom'
import is from 'prop-types'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import {browserHistory} from 'react-router'
import $ from 'jquery'

import defaultState from '../components/AddArticlePage/defaultState'
import { controlModal, getPublications, editForm, deleteCard, clearForm, cartUpdate } from '../actions/application'
import {finishUpdate} from '../utilities/helpers'
import {cardNamesArray} from '../utilities/crossmarkHelpers'
import AddArticleView from '../components/AddArticlePage/addArticleView'
import {routes} from '../routing'
import {asyncValidateArticle} from '../utilities/validation'
import {getSubItems} from '../utilities/getSubItems'
import ReviewArticle from '../components/AddArticlePage/reviewArticleModal'
import { XMLSerializer, DOMParser } from 'xmldom'
import loadArticle from '../components/AddArticlePage/methods/loadArticle'
import save from '../components/AddArticlePage/methods/save'





const mapStateToProps = (state, props) => {
  return ({
    publication: state.publications[props.routeParams.pubDoi],
    reduxForm: state.reduxForm,
    crossmarkPrefixes: state.login['crossmark-prefixes'],
    reduxCart: state.cart
  })
}

const mapDispatchToProps = dispatch => bindActionCreators({
  reduxControlModal: controlModal,
  reduxEditForm: editForm,
  reduxDeleteCard: deleteCard,
  reduxClearForm: clearForm,
  asyncGetPublications: getPublications,
  reduxCartUpdate: cartUpdate
}, dispatch)


@connect(mapStateToProps, mapDispatchToProps)
export default class AddArticlePage extends Component {

  static propTypes = {
    reduxControlModal: is.func.isRequired,
    reduxEditForm: is.func.isRequired,
    reduxClearForm: is.func.isRequired,
    reduxDeleteCard: is.func.isRequired,
    reduxCartUpdate: is.func.isRequired,
    asyncGetPublications: is.func.isRequired,

    routeParams: is.shape({
      pubDoi: is.string.isRequired,
      articleDoi: is.string
    }).isRequired,

    location: is.shape({
      state: is.shape({
        duplicateFrom: is.string
      })
    }).isRequired,

    reduxForm: is.object.isRequired,
    publication: is.object,
    crossmarkPrefixes: is.array.isRequired,
    reduxCart: is.array.isRequired
  }

  constructor (props) {
    super(props)
    const {duplicateFrom, dupIssueDoi, dupIssueTitle} = props.location.state || {}
    const ownerPrefix = props.routeParams.pubDoi.split('/')[0];
    const editArticleDoi = props.routeParams.articleDoi || (props.location.state && props.location.state.duplicateFrom)
    const isDuplicate = !!duplicateFrom
    const issueId = props.routeParams.issueId || dupIssueDoi || dupIssueTitle
    const issueDoi = issueId && (issueId.split('/')[0] === ownerPrefix) ? issueId : undefined
    const issueTitle = issueId && !issueDoi ? JSON.parse(issueId) : undefined
    this.state = {
      ...defaultState,
      publication: props.publication,
      publicationMetaData: {},
      publicationXml: '',
      issue: undefined,
      editArticleDoi,
      issueDoi,
      issueTitle,
      mode: editArticleDoi ? 'edit' : 'add',
      isDuplicate,
      ownerPrefix,
      crossmark: props.crossmarkPrefixes.indexOf(ownerPrefix) !== -1,
      crossmarkCards: {},
      version: '1',
      errorMessages: [],
      focusedInput: '',
      activeCalendar: '',
      scrollToTopButton: false,
      scrollTimeout: null
    }
    this.state.article.doi = ownerPrefix + '/'
  }


  componentDidMount () {
    document.addEventListener('scroll', this.scrollHandler, false)
    loadArticle.call(this)
  }

  save = save.bind(this)


  componentWillReceiveProps (nextProps) {
    if (this.props.crossmarkPrefixes !== nextProps.crossmarkPrefixes) {
      this.setState({
        crossmark: nextProps.crossmarkPrefixes.indexOf(this.state.ownerPrefix) !== -1
      })
    }
  }


  validation = async (data = this.state, reduxForm = this.props.reduxForm, doiDisabled = this.state.doiDisabled) => {
    const {
      criticalErrors,
      warnings,
      licenses,
      contributors,
      relatedItems,
      newReduxForm,
    } = await asyncValidateArticle(data, reduxForm, this.state.ownerPrefix, doiDisabled)

    const validatedPayload = {
      validating: true,
      mode: 'edit',
      error: false,
      errors: {...criticalErrors, ...warnings},
      criticalErrors: criticalErrors,
      license: licenses.length ? licenses : defaultState.license,
      contributors: contributors.length ? contributors : defaultState.contributors,
      relatedItems: relatedItems.length ? relatedItems : defaultState.relatedItems,
      openItems: {
        Contributors: !!contributors.length,
        Funding: !!getSubItems(data.funding).length,
        Licenses: !!licenses.length || !!data.article.freetolicense,
        relatedItems: !!relatedItems.length,
        addInfo: !!getSubItems(data.addInfo).length,
        references: !!data.references.length
      }
    }
    let valid = true

    for(const key in warnings) {
      if (warnings[key]) {
        validatedPayload.error = true
      }
    }

    for(const key in criticalErrors) {
      if(criticalErrors[key]) {
        validatedPayload.error = true
        valid = false
      }
    }

    validatedPayload.errorMessages = this.errorUtility.onValidate(validatedPayload.errors, validatedPayload.contributors, validatedPayload.license, validatedPayload.relatedItems, newReduxForm)

    if(newReduxForm && newReduxForm.size) {
      const keyPath = []
      this.props.reduxEditForm(keyPath, newReduxForm)
    }

    return {valid, validatedPayload}
  }


  validate = async () => {
    if(this.state.mode === 'edit') {
      const {validatedPayload} = await this.validation()
      this.setState(validatedPayload)
    }
  }


  errorUtility = {
    errorIndicators: [],

    activeIndicator: -1,

    openingSubItem: false,
    subItemIndex: "0",

    stickyErrorMounted: false,

    saveRef: (activeErrors, indicatorErrors, node, subItem, subItemIndex, openSubItem) => {
      if(node){
        this.errorUtility.errorIndicators.push({
          activeErrors,
          indicatorErrors,
          node,
          subItem,
          subItemIndex,
          openSubItem
        })

        if(node.id === 'errorBubble') {
          this.errorUtility.activeIndicator = this.errorUtility.errorIndicators.length - 1
        }
      }
    },

    setErrorMessages: (setErrors, allErrors = this.state.errors) => {
      const filteredErrorMessage = setErrors.filter((error)=>{
        return allErrors[error]
      })

      this.setState({errorMessages: filteredErrorMessage})
    },

    onValidate: (newValidationErrors, contributors, license, relatedItems, newReduxForm) => {
      if(!this.state.errorMessages.length) {
        return []
      }

      const {errorIndicators, activeIndicator} = this.errorUtility
      const activeIndicatorObj = errorIndicators[activeIndicator]
      const trackedIndicatorErrors = activeIndicatorObj ? activeIndicatorObj.indicatorErrors : []

      let newErrorMessages
      const {subItem, subItemIndex} = activeIndicatorObj || {}

      if(subItem) {
        const subItemErrors = {
          contributor: contributors,
          license: license,
          relatedItems: relatedItems,
        }

        let allErrors
        if(cardNamesArray.indexOf(subItem) > -1) {
          //Is crossmark subItem
          allErrors = newReduxForm.getIn([subItem, subItemIndex, 'errors'])

        } else {
          allErrors = subItemErrors[subItem][subItemIndex].errors
        }

        newErrorMessages = this.state.errorMessages.filter((error) => {
          return allErrors[error]
        })

        if(!newErrorMessages.length) {
          newErrorMessages = trackedIndicatorErrors.filter((error) => {
            return allErrors[error]
          })
        }


      } else {
        newErrorMessages = this.state.errorMessages.filter((error) => {
          return newValidationErrors[error]
        })

        if(!newErrorMessages.length) {
          newErrorMessages = trackedIndicatorErrors.filter((error)=>{
            return newValidationErrors[error]
          })
        }

        this.errorUtility.subItemIndex = "0"
      }

      if(!newErrorMessages.length) {
        const indicatorBelow = errorIndicators[activeIndicator + 1]
        const indicatorAbove = errorIndicators[activeIndicator - 1]

        if(indicatorBelow) {
          newErrorMessages = indicatorBelow.activeErrors
          this.errorUtility.subItemIndex = indicatorBelow.subItemIndex || "0"
        } else if(indicatorAbove) {
          newErrorMessages = indicatorAbove.activeErrors
          this.errorUtility.subItemIndex = indicatorAbove.subItemIndex || "0"
        }
      }

      return newErrorMessages
    },

    assignStickyErrorRefresh: (func) => {
      this.errorUtility.stickyErrorMounted = true
      this.errorUtility.refreshTask = func
    },

    refreshStickyError: () => {
      if(this.errorUtility.stickyErrorMounted && typeof this.errorUtility.refreshTask === 'function') {
        return finishUpdate().then(()=>this.errorUtility.refreshTask())
      }
    }
  }


  tooltipUtility = {

    getFocusedInput: () => this.state.focusedInput,

    tooltipMounted: false,

    assignRefreshTask: (func) => {
      this.tooltipUtility.tooltipMounted = true
      this.tooltipUtility.refreshTask = func
    },

    refresh: (param) => {
      if(
        this.tooltipUtility.tooltipMounted &&
        typeof this.tooltipUtility.refreshTask === 'function'
      ) {
        return finishUpdate().then(()=>this.tooltipUtility.refreshTask(param))
      }
    },

    assignFocus: (inputId, tooltip) => {
      this.setState({focusedInput: inputId})
      return this.tooltipUtility.refresh(tooltip)
    }
  }


  componentDidUpdate(prevProps, prevState) {
    //Select first error if first validation
    if(
      this.state.validating &&
      this.state.error && !prevState.error
    ) {
      try {
        this.errorUtility.setErrorMessages(this.errorUtility.errorIndicators[0].activeErrors)
      } catch (e) {}
    }

    this.errorUtility.refreshStickyError()

    this.tooltipUtility.refresh()

    this.state.validating = false
    this.state.saving = false
    this.state.openSubItems = false
  }


  openReviewArticleModal = () => {
    this.props.reduxControlModal({
      showModal: true,
      title:
        <div className='innerTitleHolder'>
          <div className='innterTitleHolderIcon'>
            <img src={`${routes.images}/ReviewArticle/Asset_Icons_White_Review.svg`} />
          </div>
          <div className='innerTitleHolderText'>
            {this.state.article.title}
          </div>
        </div>,
      style: 'defaultModal reviewModal',
      Component: ReviewArticle,
      props: {
        submit: this.addToCart,
        reviewData: this.state,
        publication: this.state.publication,
        publicationMetaData: this.state.publicationMetaData,
        issue: this.state.issue
      }
    })
  }


  handleChange = async (e) => {
    this.setState({
      article: {
        ...this.state.article,
        [e.target.name]: e.target.value
      }
    })
  }


  boundSetState = (...args) => { this.setState(...args) }


  toggleFields = () => {
    this.setState({
      showOptionalTitleData: !this.state.showOptionalTitleData
    })
  }


  addSection = (section) => {
    this.setState({
      [section]: [...this.state[section], defaultState[section][0]]
    })
  }


  removeSection = (section, index) => {
    this.setState({
      [section]: do {
        const newArray = [...this.state[section]]
        newArray.splice(index, 1)
        newArray
      }
    })
  }


  addToCart = () => {
    const addToCart = true
    this.save(addToCart)
  }


  back = () => {
    browserHistory.push(`${routes.publications}/${encodeURIComponent(this.state.publication.message.doi)}`)
  }


  crossmarkUtility = {

    addCrossmarkCard: (selection) => {
      this.setState({
        crossmarkCards: {
          ...this.state.crossmarkCards,
          [selection]: this.state.crossmarkCards[selection] ? this.state.crossmarkCards[selection] + 1 : 1
        }
      })
    },

    removeCrossmarkCard: (selection) => {
      const newState = {...this.state.crossmarkCards}
      delete newState[selection]
      this.props.reduxDeleteCard([selection])
      this.setState({
        crossmarkCards: newState
      })
    }
  }


  componentWillUnmount () {
    document.removeEventListener('scroll', this.scrollHandler, false)
    this.props.reduxClearForm();
    clearTimeout(this.state.scrollTimer)
  }


  calendarHandler = (name, dateObj, subItem = 'article', subItemIndex) => {
    if(dateObj) {
      const stateName = this.state.activeCalendar.split('-')[0]
      const datePayload = dateObj ? {
        [`${stateName}Year`]: dateObj.year,
        [`${stateName}Month`]: dateObj.month,
        [`${stateName}Day`]: dateObj.day
      } : {}

      const subItemPayload = subItemIndex ? this.state[subItem].map((item, i) => {
        if(i === Number(subItemIndex)) {
          return {
            ...item,
            ...datePayload
          }
        } else {
          return item
        }
      })
        : {...this.state.article, ...datePayload}

      this.setState({
        activeCalendar: name,
        [subItem]: subItemPayload
      }, () => this.validate())

    } else {
      this.setState({activeCalendar: name})
    }
  }


  scrollHandler = () => {
    if(this.state.scrollTimer) {
      clearTimeout(this.state.scrollTimer)
    }
    this.state.scrollTimer = setTimeout(() => {
      if(window.pageYOffset > 240 && !this.state.scrollToTopButton) {
        this.setState({scrollToTopButton: true})
      }

      if(window.pageYOffset < 240 && this.state.scrollToTopButton) {
        this.setState({scrollToTopButton: false})
      }
    }, 80)
  }


  scrollToTop = () => {
    $('html, body').animate({
      scrollTop: 0
    }, 1000);
  }


  render () {
    this.errorUtility.errorIndicators = []  //Saving refs of any errorIndicators rendered so need to clear it before each render
    this.errorUtility.openingSubItem = false
    return (
      <div className='addArticles'>

        <AddArticleView
          back={this.back}
          addToCart={this.addToCart}
          validate={this.validate}
          save={this.save}
          openReviewArticleModal={this.openReviewArticleModal}
          handleChange={this.handleChange}
          toggleFields={this.toggleFields}
          boundSetState={this.boundSetState}
          removeSection={this.removeSection}
          addSection={this.addSection}
          reduxDeleteCard={this.props.reduxDeleteCard}
          reduxForm={this.props.reduxForm}
          errorUtility={this.errorUtility}
          crossmarkUtility={this.crossmarkUtility}
          tooltipUtility={this.tooltipUtility}
          calendarHandler={this.calendarHandler}
          {...this.state}
        />

        <div className="rightBar">
          {this.state.scrollToTopButton &&
            <div className="scrollToTopButton" onClick={this.scrollToTop}>
              <img className="scrollToTopChevron" src={`${routes.images}/AddArticle/Asset_Icons_White_Chevron.svg`}/>
            </div>}
        </div>
      </div>
    )
  }
}

