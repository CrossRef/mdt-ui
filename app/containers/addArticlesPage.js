import React, { Component } from 'react'
import is from 'prop-types'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { Link, browserHistory } from 'react-router'
import { stateTrackerII } from 'my_decorators'

import { controlModal, getPublications, editForm, clearForm, submitArticle, cartUpdate, getItem } from '../actions/application'
import xmldoc from '../utilities/xmldoc'
import AddArticleCard from '../components/addArticleCard'



const mapStateToProps = (state, props) => {
  return ({
    publication: state.publications[props.routeParams.doi],
    reduxForm: state.reduxForm,
    crossmarkPrefixes: state.login['crossmark-prefixes'],
    reduxCart: state.cart
  })
}

const mapDispatchToProps = dispatch => bindActionCreators({
  reduxControlModal: controlModal,
  reduxEditForm: editForm,
  reduxClearForm: clearForm,
  asyncGetItem: getItem,
  asyncGetPublications: getPublications,
  asyncSubmitArticle: submitArticle,
  reduxCartUpdate: cartUpdate
}, dispatch)


@connect(mapStateToProps, mapDispatchToProps)
export default class AddArticlesPage extends Component {

  static propTypes = {
    reduxControlModal: is.func.isRequired,
    reduxEditForm: is.func.isRequired,
    reduxClearForm: is.func.isRequired,
    reduxCartUpdate: is.func.isRequired,
    asyncGetPublications: is.func.isRequired,
    asyncSubmitArticle: is.func.isRequired,
    asyncGetItem: is.func.isRequired,
    reduxCart: is.array.isRequired,
    routeParams: is.shape({
      pubDoi: is.string.isRequired,
      articleDoi: is.string
    }).isRequired,
    location: is.shape({
      state: is.shape({
        duplicateFrom: is.string
      })
    }).isRequired,
    reduxForm: is.object.isRequired,
    publication: is.object,
    crossmarkPrefixes: is.array.isRequired
  }

  constructor (props) {
    super(props)
    this.state = {
      publication: {},
      publicationMetaData: {},
      issuePublication: undefined,
      mode: 'add'
    }
  }

  componentDidMount () {
    const { duplicateFrom } = this.props.location.state || {};
    const { pubDoi, articleDoi, issueDoi } = this.props.routeParams;

    const dois = [articleDoi || duplicateFrom || pubDoi];

    if(issueDoi) dois.push(issueDoi)
    if(articleDoi || duplicateFrom) dois.push(pubDoi)

    this.props.asyncGetPublications( dois, (publications) => {
      this.setState({
        issuePublication: publications[0]
      })

      var publMeta = publications[1] ? publications[1].message.content : undefined
      var article = publications[0]
      if (issueDoi) {
        publMeta = publications[2] ? publications[2].message.content : undefined
        //doing logic here so we don't have to change the addArticles page any further
        var unwrappedPub = publications[1]

        unwrappedPub.message.contains = [publications[0].message.contains[0].contains[0]]

        article = unwrappedPub
      }

      this.setState({
        mode: (articleDoi || duplicateFrom) ? 'edit' : 'add',
        publication: article,
        publicationMetaData: publMeta ? xmldoc(publMeta) : {}
      })

    })

  }

  componentWillUnmount () {
    this.props.reduxClearForm();
  }

  render () {
    const { publication, mode, publicationMetaData } = this.state
    const { pubDoi, doi, issueDoi } = this.props.routeParams
    const ownerPrefix = this.props.publication ? this.props.publication.message['owner-prefix'] : pubDoi.split('/')[0];
    return (
      <div className='addArticles'>
        <AddArticleCard
          reduxControlModal = {this.props.reduxControlModal}
          reduxEditForm={this.props.reduxEditForm}
          reduxCartUpdate={this.props.reduxCartUpdate}
          asyncSubmitArticle={this.props.asyncSubmitArticle}
          asyncGetItem={this.props.asyncGetItem}
          reduxForm={this.props.reduxForm}
          ownerPrefix={ownerPrefix}
          publication = { publication }
          publicationMetaData = { publicationMetaData }
          issuePublication = {this.state.issuePublication}
          mode = { mode }
          issue = { issueDoi }
          isDuplicate = { this.props.location.state ? !!this.props.location.state.duplicateFrom : false }
          crossmarkPrefixes = { this.props.crossmarkPrefixes }
          reduxCart = {this.props.reduxCart}
        />
      </div>
    )
  }
}

