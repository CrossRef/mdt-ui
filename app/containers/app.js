import React, { Component } from 'react'
import {bindActionCreators} from 'redux'
import is from 'prop-types'
import { connect } from 'react-redux'
import { stateTrackerII } from 'my_decorators'

import Header from '../components/header'
import Footer from '../components/footer'
import Modal from './modal'
import {clearCartToast} from '../actions/application'


const mapStateToProps = (state) => {
  return {
    cart: state.cart,
    cartToast: state.cartToast
  }
}

const mapDispatchToProps = (dispatch) => bindActionCreators({
  reduxClearCartToast: clearCartToast
}, dispatch)

@connect(mapStateToProps, mapDispatchToProps)
export default class App extends Component {

  static propTypes = {}

  render () {
    return (
      <div className='app'>
        <div className='app-contents'>
          <Header cart={this.props.cart} cartToast={this.props.cartToast} reduxClearCartToast={this.props.reduxClearCartToast}/>
          <div className='page-contents'>
            {this.props.children}
          </div>
        </div>
        <Footer />
        <Modal />
      </div>
    )
  }
}
