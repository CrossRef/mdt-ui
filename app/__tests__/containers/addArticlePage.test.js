import React from 'react';
import Enzyme, { shallow, mount, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });
import toJson from 'enzyme-to-json';
import sinon from 'sinon'


import {AddArticlePage} from '../../containers/addArticlePage'
import defaultArticle from '../../__testUtils/dummyRecords/defaultArticle'
import fullFormArticle from '../../__testUtils/dummyRecords/fullFormArticle'


jest.mock("../../actions/api")


const didMount = sinon.spy(AddArticlePage.prototype, 'componentDidMount')
const didUpdate = sinon.spy(AddArticlePage.prototype, 'componentDidUpdate')


const commonProps = {
  reduxControlModal: () => {},
  reduxEditForm: () => {},
  reduxClearForm: () => {},
  reduxDeleteCard: () => {},
  reduxCartUpdate: () => {},
  asyncGetPublications: () => {},
  reduxForm: {},
  reduxCart: [],
  showSection: {},
  crossmarkPrefixes: []
}


function testArticle ({pubDoi, editArticleDoi, duplicateArticleDoi}) {
  let component
  const articleProps = {
    ...commonProps,
    routeParams: {pubDoi: pubDoi},
    location: {}
  }
  if(editArticleDoi) {
    articleProps.routeParams.articleDoi = editArticleDoi
  }

  if(duplicateArticleDoi) {
    articleProps.location = {state: {duplicateFrom: duplicateArticleDoi}}
  }

  beforeEach(() => {
    didMount.reset()
    didUpdate.reset()

    component = shallow(
      <AddArticlePage {...articleProps}/>
    )
  })

  it('article loads', () => {
    expect(didMount.callCount).toBe(1)
    expect(didUpdate.called).toBe(true)
    expect(toJson(component)).toMatchSnapshot()
  })

}





describe('edit defaultArticle', ()=>testArticle({pubDoi: defaultArticle.publicationDoi, editArticleDoi: defaultArticle.articleDoi}))



describe('edit fullFormArticle', ()=>testArticle({pubDoi: fullFormArticle.publicationDoi, editArticleDoi: fullFormArticle.articleDoi}))



describe('duplicate article tests', ()=>testArticle({pubDoi: defaultArticle.publicationDoi, duplicateArticleDoi: defaultArticle.articleDoi}))



describe('new article tests', ()=>testArticle({pubDoi: defaultArticle.publicationDoi}))



describe('save defaultArticle', () => {
  let component
  const articleProps = {
    ...commonProps,
    routeParams: {pubDoi: defaultArticle.publicationDoi, articleDoi: defaultArticle.articleDoi},
    location: {}
  }

  beforeEach(() => {
    didMount.reset()
    didUpdate.reset()

    component = shallow(
      <AddArticlePage {...articleProps}/>
    )

    component.setState({})
  })

  it('save', () => {
    expect(didMount.callCount).toBe(1)
    expect(didUpdate.called).toBe(true)
    expect(toJson(component)).toMatchSnapshot()
  })


})



// it('should update', () => {
//   component.setProps({
//     criticalErrors: {
//       title: true,
//       doi: true,
//       invaliddoi: true,
//       invalidDoiPrefix: true,
//       freetolicense: true,
//       dupedoi: true
//     },
//     saving: true
//   })
//
//   expect(toJson(component)).toMatchSnapshot()
//   expect(didMount.calledTwice).toBe(false)
//   expect(didReceiveProps.calledOnce).toBe(true)
// })
