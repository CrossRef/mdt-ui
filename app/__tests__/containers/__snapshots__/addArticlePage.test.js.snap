// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`save defaultArticle save 1`] = `
<div
  className="addArticles"
>
  <AddArticleView
    activeCalendar=""
    addInfo={
      Object {
        "archiveLocation": "",
        "language": "",
        "publicationType": "",
        "similarityCheckURL": "http://",
      }
    }
    addSection={[Function]}
    addToCart={[Function]}
    article={
      Object {
        "abstract": "",
        "doi": "10.5555/",
        "firstPage": "",
        "freetolicense": "",
        "lastPage": "",
        "locationId": "",
        "onlineDateDay": "",
        "onlineDateMonth": "",
        "onlineDateYear": "",
        "originallanguagetitle": "",
        "originallanguagetitlesubtitle": "",
        "printDateDay": "",
        "printDateMonth": "",
        "printDateYear": "",
        "subtitle": "",
        "title": "",
        "url": "http://",
      }
    }
    back={[Function]}
    boundSetState={[Function]}
    calendarHandler={[Function]}
    contributors={
      Array [
        Object {
          "affiliation": "",
          "errors": Object {
            "contributorGroupName": false,
            "contributorGroupRole": false,
            "contributorLastName": false,
            "contributorRole": false,
          },
          "firstName": "",
          "groupAuthorName": "",
          "groupAuthorRole": "",
          "lastName": "",
          "orcid": "",
          "role": "",
          "suffix": "",
        },
      ]
    }
    criticalErrors={Object {}}
    crossmark={false}
    crossmarkCards={Object {}}
    crossmarkUtility={
      Object {
        "addCrossmarkCard": [Function],
        "removeCrossmarkCard": [Function],
      }
    }
    doiDisabled={false}
    editArticleDoi="10.5555/defaultArticle"
    error={false}
    errorMessages={Array []}
    errorUtility={
      Object {
        "activeIndicator": -1,
        "assignStickyErrorRefresh": [Function],
        "errorIndicators": Array [],
        "onValidate": [Function],
        "openingSubItem": false,
        "refreshStickyError": [Function],
        "saveRef": [Function],
        "setErrorMessages": [Function],
        "stickyErrorMounted": false,
        "subItemIndex": "0",
      }
    }
    errors={
      Object {
        "contributorGroupName": false,
        "contributorGroupRole": false,
        "contributorLastName": false,
        "contributorRole": false,
        "doi": false,
        "dupedoi": false,
        "firstPage": false,
        "invalidDoiPrefix": false,
        "invaliddoi": false,
        "invalidurl": false,
        "licenseDateIncomplete": false,
        "licenseDateInvalid": false,
        "licenseUrl": false,
        "licenseUrlInvalid": false,
        "onlineDateIncomplete": false,
        "onlineDateInvalid": false,
        "onlineDateYear": false,
        "printDateIncomplete": false,
        "printDateInvalid": false,
        "printDateYear": false,
        "relatedItemDoiInvalid": false,
        "relatedItemIdType": false,
        "relatedItemRelType": false,
        "simCheckUrlInvalid": false,
        "title": false,
        "url": false,
      }
    }
    focusedInput=""
    funding={
      Array [
        Object {
          "funderName": "",
          "funder_identifier": "",
          "grantNumbers": Array [
            "",
          ],
        },
      ]
    }
    handleChange={[Function]}
    isDuplicate={false}
    license={
      Array [
        Object {
          "acceptedDateDay": "",
          "acceptedDateMonth": "",
          "acceptedDateYear": "",
          "appliesto": "",
          "errors": Object {
            "freetolicense": false,
            "licenseDateIncomplete": false,
            "licenseDateInvalid": false,
            "licenseUrl": false,
            "licenseUrlInvalid": false,
          },
          "licenseurl": "http://",
        },
      ]
    }
    mode="edit"
    openItems={
      Object {
        "Contributors": false,
        "Funding": false,
        "Licenses": false,
        "References": false,
        "RelatedItems": false,
        "addInfo": false,
      }
    }
    openReviewArticleModal={[Function]}
    ownerPrefix="10.5555"
    publicationMetaData={Object {}}
    publicationXml=""
    reduxDeleteCard={[Function]}
    reduxForm={Object {}}
    references={Array []}
    relatedItems={
      Array [
        Object {
          "description": "",
          "errors": Object {
            "relatedItemDoiInvalid": false,
            "relatedItemIdType": false,
            "relatedItemRelType": false,
          },
          "identifierType": "",
          "relatedItemIdentifier": "",
          "relationType": "",
        },
      ]
    }
    removeSection={[Function]}
    save={[Function]}
    saving={false}
    scrollTimeout={null}
    scrollToTopButton={false}
    showAdditionalInformation={false}
    showCards={Object {}}
    showContributor={false}
    showFunding={false}
    showHelper={true}
    showOptionalTitleData={false}
    showRelatedItems={false}
    toggleFields={[Function]}
    tooltipUtility={
      Object {
        "assignFocus": [Function],
        "assignRefreshTask": [Function],
        "getFocusedInput": [Function],
        "refresh": [Function],
        "tooltipMounted": false,
      }
    }
    validate={[Function]}
    validating={false}
    version="1"
  />
  <div
    className="rightBar"
  />
</div>
`;
